cmd/stats/stats.go

```diff
@@ -185,19 +197,185 @@ func argsToColumnFilter(cmd *cobra.Command) []string {
+		Name string
+		// "since X days", or "from YYYY-MM-DD to YYYY-MM-DD", etc
+		Date string
+	}{
+		Name: nameSection,
+		Date: dateSection,
+	}
+
+	tmplString := "" +
+		`stats {{if .Name}}for {{.Name}} {{end}}{{.Date}}:`
+
+	tmpl := template.Must(template.New("header").Parse(tmplString))
+	var buf bytes.Buffer
+	err := tmpl.Execute(&buf, tmplData)
+	if err != nil {
+		panic(err)
+	}
+	return buf.String()
+}
+
+// Calculates the average, min, max, first, last, and delta of a stat.
+// Returns that information in an array of strings.
+// Formats the data based on the column name selected.
+func getColumnStats(reps []Rep, colName string) []string {
+	getCol := func(r Rep, colName string) float64 {
+		switch colName {
+		case WPM:
+			return r.Wpm
+		case RAW_WPM:
+			return r.Raw
+		case DURATION:
+			return float64(r.Dur)
+		case ACCURACY:
+			return r.Acc
+		case MISTAKES:
+			return float64(r.Miss)
+		case UNCORRECTED_ERRORS:
+			return float64(r.Errs)
+		}
+		return -math.MaxFloat64
+	}
+
+	avg := 0.0
+	min := math.MaxFloat64
+	max := -math.MaxFloat64
+	first := getCol(reps[len(reps)-1], colName)
+	last := getCol(reps[0], colName)
+	delta := last - first
+
+	sum := 0.0
+	for _, rep := range reps {
+		curr := getCol(rep, colName)
+		sum += curr
+		if curr < min {
+			min = curr
+		}
+		if curr > max {
+			max = curr
+		}
+	}
+	avg = sum / float64(len(reps))
+
+	colData := []float64{avg, min, max, first, last, delta}
+
+	row := []string{}
+
+	for _, d := range colData {
+		row = append(row, util.ColumnString(colName, d))
+	}
+	return row
+}
+
+func printStats(cmd *cobra.Command, reps []Rep) {
+	name := cmd.Flag(NAME).Value.String()
+	lang := cmd.Flag(LANGUAGE).Value.String()
+	start := cmd.Flag(START).Value.String()
+	if since := cmd.Flag("since").Value.String(); since != "" {
+		start = since
+	}
+	end := cmd.Flag(END).Value.String()
+
 	// print the header
-	fmt.Printf("%s\n", strings.Join(cols, "\t"))
+	fmt.Println(getStatsHeader(name, lang, start, end))
+
+	cols := argsToColumnFilter(cmd)
+
+	if len(reps) == 0 {
+		fmt.Println("no stats")
+	} else if len(reps) > 1 {
+
+		// print the stats table
+		table := tw.NewWriter(os.Stdout)
+		table.SetHeader([]string{"", "avg", "min", "max", "first", "last", "delta"})
+		table.SetAutoWrapText(false)
+		table.SetAutoFormatHeaders(true)
+		table.SetHeaderAlignment(tw.ALIGN_LEFT)
+		table.SetAlignment(tw.ALIGN_LEFT)
+		table.SetCenterSeparator("")
+		table.SetColumnSeparator("")
+		table.SetRowSeparator("")
+		table.SetHeaderLine(false)
+		table.SetTablePadding("  ") // pad with tabs
+		table.SetNoWhiteSpace(true)
+		for _, col := range cols {
+			if col == START || col == NAME {
+				continue
+			}
+			row := []string{col}
+			row = append(row, getColumnStats(reps, col)...)
+			table.Append(row)
+		}
+		table.Render()
+
+		wpmGraphData := []float64{}
+		rawWpmGraphData := []float64{}
+		mistakesGraphData := []float64{}
+		accuracyGraphData := []float64{}
+		errorsGraphData := []float64{}
```

Comment:
Need to control the lines that are visible based on the arguments that are provided to the command

Reply:
I also really want to show the columns in the order in which they were added in the command line, so I need to write a function that handles that. Perhaps I can use that function to control the order and the visibility of the columns that appear as well, or modify a function that I wrote previously.